### PR TITLE
Multiple Kowalski instances

### PR DESCRIPTION
**Pull request by Theodlz**
This PR is heavily inspired by @bfhealy 's PR: https://github.com/dmitryduev/penquins/pull/20

Here, we try to make it simpler from the client side to add instances.

MISSING: What do we do when querying catalogs that are not in just one instance in one query? Do we return an error do we generate N queries to query each instance for the catalogs they have? And especially, how do we handle that problem when running batch queries?

ISSUE: What do we do when 2 instances have a catalog with the same name and the user did not specify as specific instance to query? Do we return an error, or default to the first one in the list of instance and display a warning?

### CODE REVIEW

**Comment by Theodlz** [comment]
@bfhealy @mcoughlin this is not finished at all yet, but you might want to look at it.

**Comment by bfhealy** [comment]
@Theodlz Thanks for reworking the original PR! I think running only one query (or one batch of queries) is appropriate when catalogs with the same name exist in more than one instance. The catalog names are something we have good control over, and the same catalog name should imply the same data within.

I think displaying a warning and defaulting to querying the first instance of a catalog is an acceptable way to proceed. The user should still be able to specify which instance to query if desired, but the code will be most useful if the user can generally abstract out the specific instances and what they contain.

I'll checkout the PR code to see how it runs!

File: penquins/penquins.py [hunk 1]
```diff
@@ -408,25 +605,153 @@ def query_skymap(
...
+                        list(instances_having_catalog.values()).index(True)
+                    ]
+                    if queries[instance_name] is None:
+                        # we set it to a copy of the original query, but with only the current catalog
+                        queries[instance_name] = deepcopy(query)
+                        # if it has a single "catalog" key, we remove it as we will replace it with a "catalogs" key
+                        if "catalog" in queries[instance_name]["query"]:
+                            del queries[instance_name]["query"]["catalog"]
+                        queries[instance_name]["query"]["catalogs"] = {
+                            catalog: query["query"]["catalogs"][catalog]
```
Comment (bfhealy) [inline penquins/penquins.py:729]: Hi @Theodlz - I'm getting an error:
```
File ~/penquins/penquins/penquins.py:729, in Kowalski.prepare_query(self, query, name)
    726     if "catalog" in queries[instance_name]["query"]:
    727         del queries[instance_name]["query"]["catalog"]
    728     queries[instance_name]["query"]["catalogs"] = {
--> 729         catalog: query["query"]["catalogs"][catalog]
    730     }
    731 else:
    732     # if the instance already has a query, we add the current catalog to the list of catalogs
    733     queries[instance_name]["query"]["catalogs"][catalog] = query[
    734         "query"
    735     ]["catalogs"][catalog]

KeyError: 'catalogs'
```

when I run the following query:
```
catalog = 'Gaia_EDR3'

query = {
    "query_type": "find_one",
    "query": {
        "catalog": catalog,
        "filter": {},
    
    },
    "projection":{'_id':1},
    "kwargs":{
        "max_time_ms": 10000,
    }
}

q = kowalski.query(query)
```
where `kowalski` is initialized for all three machines.

Comment (Theodlz) [reply penquins/penquins.py:730]: Hi again @bfhealy. I pushed some changes quickly (haven't tested them yet, given that the current unit tests were already passing. I need to design a test that can fail if this is wrong), maybe you can retry your query?

Comment (bfhealy) [reply penquins/penquins.py:730]: @Theodlz Running the same query as before, I'm seeing the following error:
```
File ~/penquins/penquins/penquins.py:687, in Kowalski.prepare_query(self, query, name)
    685     catalogs = query["query"]["catalogs"]
    686 else:
--> 687     catalogs = {query["query"]["catalog"]: query["query"]["projection"]}
    689 queries = {name: None for name in self.instances.keys()}
    690 if name is None:
    691     # if no name is specified, but we have only one instance, we use that instance

KeyError: 'projection'
```

Comment (Theodlz) [reply penquins/penquins.py]: Ok I will add some code later today to fix the query automatically if the user puts the projection in the wrong place. For the error below, can you retry now with the changes I just pushed?

Comment (bfhealy) [reply penquins/penquins.py]: Thanks @Theodlz, I just ran this query:
```
catalog = 'Gaia_EDR3'

query = {
    "query_type": "find_one",
    "query": {
        "catalog": catalog,
        "filter": {},
        "projection": {'_id':1}
    },
    "kwargs":{
        "max_time_ms": 10000,
    }
}

q = kowalski.query(query)
```
and got this error:
```
File ~/penquins/penquins/penquins.py:472, in Kowalski.query(self, query, queries, name, use_batch_query, max_n_threads)
    470 else:
    471     results = {}
--> 472     for name, query in queries.items():
    473         results.update(self.single_query(query, name=name))
    474     return results

AttributeError: 'NoneType' object has no attribute 'items'
```

**Comment by Theodlz** [comment]
Hi @bfhealy. I made some changes, fixed some issues and added some better testing. Could you retry with the current code? I based one test on your query so that should be fine, and I also made sure that a cone search (where you can specify multiple catalogs) also works across multiple instances.

If that works for you, then the next steps in that PR are:
- Cleaning up the code, reordering the tests. I believe it can be simplified significantly.
- When running multiple queries and a catalog is available on multiple instances, split the load for a given catalog across the instances, as the current behavior is to default to the first instance that has the catalog. I believe that this would help you a lot for your feature generation once we make copies of catalogs available on multiple instances.

Let me know what you think.

**Comment by bfhealy** [comment]
Hi @Theodlz, these changes are great and will significantly simplify querying across multiple instances! I tested out the previous query and a cone search across multiple catalogs/instances, and they both perform as I would expect them to. 

Sounds like a very good idea to split the instance load automatically - this will definitely be useful once the latest ZTF DR gets uploaded to a second instance.

**Comment by Theodlz** [comment]
Great! I will add some code to that same PR to split the load based on catalog availability across instances. I might ping you again once that goes in.

**Comment by bfhealy** [comment]
Thanks!

**Comment by Theodlz** [comment]
Hi again Brian. I added some code that seems to split the load between instances decently. So now you should be able to write a bunch of queries, and run them using:

 `kowalski.query(queries=[<your_list>], use_batch_query=True/False, max_n_threads=<your_number>)`.

Also, it will use less threads if the max nb specified doesn't need to be met, which is a nice addition.

**Comment by bfhealy** [comment]
Hi Theo, I'm still working on testing the new load-splitting code within the feature generation script, but in the mean time I encountered an error with a `catalog_names` query that was previously running:

Query:
```
{
  'query_type': 'info',
  'query': {
        'command': 'catalog_names'
        } 
}
```

Call to `penquins.Kowalski.query`:
```
kowalski.query(query, name='kowalski')
```

Error:
```
File ~/penquins/penquins/penquins.py:463, in Kowalski.query(self, query, queries, name, use_batch_query, max_n_threads)
    461     raise ValueError("Please specify either query or queries, not both")
    462 if query is not None:
--> 463     query_split_in_queries, _ = self.prepare_query(query, name=name)
    465     if name is not None:
    466         return self.single_query(query_split_in_queries[name], name=name)

ValueError: not enough values to unpack (expected 2, got 1)
```

**Comment by Theodlz** [comment]
Fixed @bfhealy, you can retry now. Thanks for noticing this one. Penquins doesn't have any tests for these kinds of "system" queries. I'm gonna add one now based on yours!

**Comment by Theodlz** [comment]
The return is fixed but it might still error when it shouldn't. On it

**Comment by Theodlz** [comment]
@bfhealy done! Let me know how it goes now. And thanks for your help figuring out those edge cases, it helps a lot to make more robust tests.

**Comment by bfhealy** [comment]
Thanks @Theodlz! Happy to help - the `catalog_names` query works fine now. I am running into another error when testing the batch query functionality within my feature generation script. I'm passing in a list of queries generated as follows:
```
queries = [
                {
                    "query_type": "find",
                    "query": {
                        "catalog": catalog,
                        "filter": {
                            "_id": {
                                "$in": 
                                    ids[
                                        i
                                        * limit_per_query : (i + 1)
                                        * limit_per_query
                                    ]
                            },
                            "data.programid": {
                                "$in": 
                                    program_id_selector,
                            }
                        },
                        "projection": {
                            "_id": 1,
                            "ra": 1,
                            "dec": 1,
                            "filter": 1,
                            "meanmag": 1,
                            "vonneumannratio": 1,
                            "refchi": 1,
                            "refmag": 1,
                            "refmagerr": 1,
                            "iqr": 1,
                            "data": 1,
                        },
                    },
                }
                for i in range(itr, itr + min(Nqueries, Ncore))
            ]
```

running this line:
```
responses = kowalski_instances.query(queries=queries, use_batch_query=True, max_n_threads=Ncore)
```

and getting the following errors:
```
Traceback (most recent call last):
  File "/Users/bhealy/penquins/../scope/tools/generate_features.py", line 874, in <module>
    generate_features(
  File "/Users/bhealy/penquins/../scope/tools/generate_features.py", line 367, in generate_features
    lcs = get_lightcurves_via_ids(
  File "/Users/bhealy/scope/scope/fritz.py", line 309, in get_lightcurves_via_ids
    responses = kowalski_instances.query(queries=queries, use_batch_query=True, max_n_threads=Ncore)
  File "/Users/bhealy/penquins/penquins/penquins.py", line 501, in query
    results = self.batch_query(
  File "/Users/bhealy/penquins/penquins/penquins.py", line 410, in batch_query
    return pool.starmap(
  File "/Users/bhealy/miniforge3/envs/scope-env-clone2/lib/python3.10/multiprocessing/pool.py", line 375, in starmap
    return self._map_async(func, iterable, starmapstar, chunksize).get()
  File "/Users/bhealy/miniforge3/envs/scope-env-clone2/lib/python3.10/multiprocessing/pool.py", line 774, in get
    raise self._value
  File "/Users/bhealy/miniforge3/envs/scope-env-clone2/lib/python3.10/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/Users/bhealy/miniforge3/envs/scope-env-clone2/lib/python3.10/multiprocessing/pool.py", line 51, in starmapstar
    return list(itertools.starmap(args[0], args[1]))
  File "/Users/bhealy/penquins/penquins/penquins.py", line 424, in single_query
    save = _query.get("kwargs", dict()).get("save", False)
AttributeError: 'NoneType' object has no attribute 'get'
```

When I run a single query of the same form, it successfully returns the data. Something related to the list of queries is preventing it from getting passed along in the code. Let me know if I can give you more information about this error.

**Comment by Theodlz** [comment]
Hi @bfhealy, I'm on it. I do see that error on non batch queries, which is solved and will be pushed soon. However, it works just fine for me with a similar query using batch querying. Could you maybe send me by email a subset of the queries you generated so I can try that? Thanks.

**Comment by Theodlz** [comment]
I managed to recreate the error! Working on fixing it now

**Comment by bfhealy** [comment]
@Theodlz Great! Just let me know if you still end up needing an example query subset.

**Comment by bfhealy** [comment]
Thanks for the fixes @Theodlz - the previous query is now running successfully, although I'm running into an error with another list of cone search queries. That seems to be the only fundamental difference I can identify. Once again, a single cone search query out of the list runs successfully. Happy to send you the list of queries by email if that would help. Here's the associated error:

```
Traceback (most recent call last):
  File "/Users/bhealy/penquins/../scope/tools/generate_features.py", line 874, in <module>
    generate_features(
  File "/Users/bhealy/penquins/../scope/tools/generate_features.py", line 630, in generate_features
    alert_stats_dct = alertstats.get_ztf_alert_stats(
  File "/Users/bhealy/scope/tools/featureGeneration/alertstats.py", line 81, in get_ztf_alert_stats
    q = kowalski_instances.query(queries=queries, use_batch_query=True, max_n_threads=Ncore)
  File "/Users/bhealy/penquins/penquins/penquins.py", line 501, in query
    results = self.batch_query(
  File "/Users/bhealy/penquins/penquins/penquins.py", line 403, in batch_query
    with ThreadPool(processes=n_threads) as pool:
  File "/Users/bhealy/miniforge3/envs/scope-env-clone2/lib/python3.10/multiprocessing/pool.py", line 930, in __init__
    Pool.__init__(self, processes, initializer, initargs)
  File "/Users/bhealy/miniforge3/envs/scope-env-clone2/lib/python3.10/multiprocessing/pool.py", line 205, in __init__
    raise ValueError("Number of processes must be at least 1")
ValueError: Number of processes must be at least 1
```

**Comment by bfhealy** [comment]
Just got a different error on that query after trying your latest push:
```
Traceback (most recent call last):
  File "/Users/bhealy/penquins/../scope/tools/generate_features.py", line 874, in <module>
    generate_features(
  File "/Users/bhealy/penquins/../scope/tools/generate_features.py", line 630, in generate_features
    alert_stats_dct = alertstats.get_ztf_alert_stats(
  File "/Users/bhealy/scope/tools/featureGeneration/alertstats.py", line 81, in get_ztf_alert_stats
    q = kowalski_instances.query(queries=queries, use_batch_query=True, max_n_threads=Ncore)
  File "/Users/bhealy/penquins/penquins/penquins.py", line 506, in query
    final_results = {
  File "/Users/bhealy/penquins/penquins/penquins.py", line 507, in <dictcomp>
    name: [result[name] for result in results if name in result]
  File "/Users/bhealy/penquins/penquins/penquins.py", line 507, in <listcomp>
    name: [result[name] for result in results if name in result]
TypeError: argument of type 'NoneType' is not iterable
```

**Comment by Theodlz** [comment]
Hmmmm, that's odd. This should happen if the list of queries is empty. Not that the list you passed is empty (I mean it could be but I don't think so), but if the list of queries they are divided in based on catalogs per instance might be. 

EDIT: I just added an error message that can help debug this. When you have some time, please retry and let me know what happens.

**Comment by bfhealy** [comment]
Interesting, I tried the query again with your latest commits and it appears to have succeeded. I'll get back at this tomorrow and keep you posted!

**Comment by bfhealy** [comment]
Hi @Theodlz, after further testing I can confirm all of my feature generation queries are working! I moved on to timing tests to see how the current and new code compare. I ran 4 tests of the feature generation script for the same ZTF quadrant of data, which produce identical results. There is a parameter `Ncore` that I pass into the script to designate how many cores to parallelize over.

For the first two tests, I set `Ncore = 1` and compared the old and new penquins code. For the run with the new code, all the Kowalski queries in my script are updated to interface with the code in this PR.
- Test 1 (old code, Ncore = 1): runtime 3:21
- Test 2 (new code, Ncore = 1): runtime 3:20

For the second two tests, I set `Ncore = 8`. I noticed that the parallelized script running on the new code took longer than the old code:
- Test 3 (old code, Ncore = 8): runtime 1:18
- Test 4 (new code, Ncore = 8): runtime 3:48

The similarity of both `Ncore = 1` runtimes seems to control for the differences in the feature generation script required to interface with the new penquins code. Maybe something in the new batch querying is slowing down the process. I'll continue looking at the changes I made in my script in case they are causing the slowdown. A typical query in the script looks like `responses = kowalski_instances.query(queries=queries, use_batch_query=True, max_n_threads=Ncore)`

**Comment by Theodlz** [comment]
Hi Brian,

That's super helpful, and I'm glad that at least it can compare when running on a single core.

For the multi core situation, here's what I suspect is happening:

I added some code to try to evaluate how many cores are actually needed, based on the number of queries generated. I guess that might be the problem. Also, I changed the method uses for parallel computing, as I now needed to pass 2 arguments to the single query method which the current multithreading method did not support. I can reverse that change pretty easily simply by passing a dict containing the 2 args in one, and we can see if that fixes it? Also, if the piece of code that is failing is the one that determines the nb of core that is really needed, you can simply add a print of the `n_threads` variable in the batch_query method right before they are executed. Maybe you could try that first and let me know if the resulting nb of core seems incorrect?

**Comment by Theodlz** [comment]
Ah and also @bfhealy. I guess that you didn't just change the Ncore but also the queries in general right? Given that the old code did not support multi instance querying, you probably query each instance using batch queries one by one right?

**Comment by bfhealy** [comment]
@Theodlz that's correct, the old code ran queries for a single Kowalski instance while the new code uses the multi-instance querying.

I found a handy commented print statement in the `batch_query` method. When I uncommented and ran my multi-core script again, it printed the correct number for the first query (`Running 8 queries on 8 threads`) but incorrect for two others (`Running 0 queries on 8 threads`). The script didn't fail though.

**Comment by Theodlz** [comment]
With the new query methods you should be able to pass all of your querues across instances so I'm not sure why you had to run it multiple times.

**Comment by bfhealy** [comment]
Sorry for the confusion, by multiple queries I mean independent ones for different data (light curves, ZTF alerts and external catalog data) at different points in the script. I could probably now consolidate the last two, but the first query for light curves fetches data for a larger list of sources that is then narrowed down in the subsequent light curve analysis.

**Comment by Theodlz** [comment]
Ok that makes sense. Then that means that whatever query you gave the method, the method didn't process it as it should and ended up returning no queries or something like that.

1. I think that this time I do need some samples of the queries you create. Could you send me that by email (tdulaz@caltech.edu)
2. Just to clarify, the 3 min vs 1 min was on all the queries or just the first ones that worked?
3. Right where you printed the nb of threads, you could also print length of the list of tuples that is defined above (which contains query + instance name), and compare it with how many queries you would have if you queried each instance individually with the old code. That might be also something that can be optimized.

**Comment by Theodlz** [comment]
Ah and interesting to know, the `starmap` method I used as a replacement to the `imap` one is `blocking`, meaning that the rest of the code has to wait for it to compute. Whereas the `imap` one is non blocking, meaning that batch query can return even when not finished, and then the result (which is an iterator) can be read like a list, and this is when the blocking part will happen if you try to access an element of the list (which is the result of one query) that isn't available yet. 

One thing about having it blocking is that I can regroup the results by instance, and have it return something like: {instance1: [query results], ... , instanceN: [query results]}, which I can't do with starmap.

I will try to reverse to imap, but expect the results to be different. Now it will be a list of dict each with the result of the query and the instance it came from.

**Comment by bfhealy** [comment]
Thanks @Theodlz, I'll try it again!

To clarify # 2 above, the 3 vs 1 minute was to run through the whole script including all the queries.

For # 3, I'm seeing 8, 8 and 16 for the length of `queries_name_tpl`. My guess on the 16 is because one or more external catalogs are hosted on more than one Kowalski instance.

**Comment by bfhealy** [comment]
Just timed the whole script at 1:02 (compared to 3:48 before)!

**Comment by Theodlz** [comment]
> Just timed the whole script at 1:02 (compared to 3:48 before)!

Awesome! Last thing to figure out is the empty list of queries that I think might be generated here which resulted in the issues we had yesterday with a ntheads that ended up at 0.

You might want to keep the commented print in the batch query uncommented to track that the n of threads is above one when it should (that is, whenever you list of queries has more than one query, or if the only query you run has catalogs across instances.

**Comment by bfhealy** [comment]
@Theodlz I'll keep that code uncommented as I continue to test. I have not been encountering `nthreads = 0` recently, but I have been seeing more of `Running 0 queries on 8 threads`. I think what's happening there is that the last instance to potentially be queried is melman, but there are no requested catalogs on that instance, so I see a `'melman': []` in the dictionary of queries passed into `batch_query`.

**Comment by Theodlz** [comment]
> @Theodlz I'll keep that code uncommented as I continue to test. I have not been encountering `nthreads = 0` recently, but I have been seeing more of `Running 0 queries on 8 threads`. I think what's happening there is that the last instance to potentially be queried is melman, but there are no requested catalogs on that instance, so I see a `'melman': []` in the dictionary of queries passed into `batch_query`.

That does match what I thought it could be. Could you send me some examples of what query type might yield no queries? I need to figure out why the `prepare_query` method might be ignoring them. There are queries in Kowalski that require catalogs and some that don't, and I probably just need to handle that difference a little better. Based on that failed query, I will as usual add a test that makes sure this we cover that possibility.

**Comment by bfhealy** [comment]
Hi @Theodlz, I think I have fixed the issue leading to empty queries in my feature generation script. The issue arose from the small number of sources being queried (10) compared to the number of cores specified (8). Everything still seems to be running smoothly, although even after this fix I continue to get an unusual output from the debugging statements, e.g.:
```
Split the 0 queries dict (instance: queries) into 16 queries tuple (instance, query)
Running 16 queries on 8 threads
```

I think the "0 queries" comes from the use of the same `queries` variable name in two places in lines 394-396 of `penquins.py`:
```
for name, queries in queries.items():
            for query in queries:
                queries_name_tpl.append((query, name))
```
In this loop, I think the `melman` instance is the last of the three instances to be iterated over. When `melman` does not contain the catalog of interest, `queries` will empty at the conclusion of this loop, and so `len(queries)` will be 0. This is the case for a query of Gaia EDR3, which is only on `kowalski` and `gloria`. When I change the code to something like
```
for name, qs in queries.items():
            for query in qs:
                queries_name_tpl.append((query, name))
```
The debug output becomes e.g.
```
Split the 3 queries dict (instance: queries) into 16 queries tuple (instance, query)
Running 16 queries on 8 threads
```
as expected (I think!).

**Comment by Theodlz** [comment]
Hi @bfhealy, I see the issue with the print. So just to be clear, that is just an issue with what is printed then, and no issues with the actual output? Anyway, this print will be removed once merged with main so that won't be a problem.

**Comment by Theodlz** [comment]
Awesome @bfhealy, thanks! I will spend some time this week cleaning up the code a bit here and there if I can, after which I will ask for a last review and merge if everything is still fine!

**Review by mcoughlin — APPROVED** [review]
LGTM!
